### PR TITLE
Shadow blocks morph to regular one on change (see #311), as an optional behavior

### PR DESCRIPTION
**Pull request by miguel76**
I added back the previous behavior of shadow blocks (they morph to regular blocks upon any change to their fields) in a way that is compatible with the undo/redo mechanism.
I created a new kind of `Blockly.Events.CHANGE` event, with `element` having value `'shadow'`, that gets called upon any call on `.setShadow()` changing the "shadow state" of the block.
When another `Blockly.Events.CHANGE` event gets called for shadow block, `.setShadow(false)` is called. The resulting new event is grouped with the current one, to ensure it can be undo/redo together.
I added the workspace option `shadowMorph` to activate this behavior, while leaving the pre-existing behavior as default.

### CODE REVIEW

**Comment by miguel76** [comment]
I put the main logic at the begin of `Blockly.Events.fire()` because it is the earliest point in which I could intercept all the `Blockly.Events.CHANGE` events.
Another possible solution (that I discarded for being much more invasive of the existing code, although possibly 'cleaner') would be to have a sort of event broker at the level of the block through which all the change events would go before reaching `Blockly.Events.fire()`. There the check would be made before calling `Blockly.Events.fire()`. But that required to change all the places in which `Blockly.Events.CHANGE` events are fired.

**Comment by RoboErikG** [comment]
If you edit a shadow (making it not a shadow) then remove it does the parent block create a new shadow again?

If not, another way to do this would be to cancel the edit event and instead create a new block with the edited value and connect it (hiding the shadow).

**Comment by miguel76** [comment]
Yes, removing the regular block (that was shadow) the shadow block is created again, based on `getShadowDom()` of the parent connection.
At first I thought also to create a new regular block, but that would interfere with field editing.

**Comment by RoboErikG** [comment]
Ignore me then. Thanks for double checking. =)

**Comment by miguel76** [comment]
Now I made the proposal cleaner and more conservative by moving the check from generic event handling code in `Blockly.Event.fire()` to `Blockly.Field.prototype.setValue()`.
Now the morphing behavior is restricted to the cases in which the value in one the fields of a shadow block is changed (in the first version of the PR, any `Blockly.Events.CHANGE` event on a shadow block changed it to a regular block).

File: core/field.js [hunk 1]
```diff
@@ -407,9 +407,30 @@ Blockly.Field.prototype.setValue = function(newText) {
...
     return;
   }
-  if (this.sourceBlock_ && Blockly.Events.isEnabled()) {
-    Blockly.Events.fire(new Blockly.Events.Change(
-        this.sourceBlock_, 'field', this.name, oldText, newText));
+  if (this.sourceBlock_) {
+    // If the shadowMorph option is enabled in the workspace, any change in the
+    // field value trigger the parent block to become "not shadow".
+    var noGroup = false;
+    if (workspace.options.shadowMorphEnabled &&
```
Comment (jano42) [inline core/field.js]: this should be 

``````js

this.sourceBlock_.workspace.options.shadowMorphEnabled

``````

**Comment by rachel-fenichel** [comment]
Hi @miguel76,

To avoid adding even more to our configuration, we're sticking with our current shadow blocks implementation.  I'm going to tag this as an enhancement and leave it open, so that anybody else who wants this behaviour can pull it in.

**Comment by pelikhan** [comment]
Can we vote to get this config flag back in?

Sent from Outlook Mobile<https://aka.ms/djubve>

________________________________

**Comment by miguel76** [comment]
Thanks to @jano42 for reporting the bug in field.js. I fixed it as suggested.

**Comment by jano42** [comment]
I've tested this PR on my branch and I've found some issues.

As a summary, the PR doesn't work for all field type.
I've extended it to all fields.


I suggest you to see what I changed when merging your PR code into my branch
https://github.com/jano42/blockly/commit/38e1e1e58806fc1ab234f83e1be5f6339739a95e

**Comment by rachel-fenichel** [comment]
Vote noted, but at this time we are not planning to support it on our end.

**Comment by reedspool** [comment]
+1 to the noted vote, this feature makes or breaks shadow blocks for our application. We've custom-implemented it before seeing this contribution, but would love to see it in core.

**Comment by BeksOmega** [comment]
Closing this as stale. New tracking issue is https://github.com/google/blockly-samples/issues/878

**Pull request closed without merging**